### PR TITLE
Refactor acorn info and add -A functionality

### PR DESCRIPTION
**Pull request by jacobdonenfeld**
### Checklist
- [x] All relevant issues are referenced in this PR
- [x] The title of this PR would make a good line in the Release Note's Changelog
- [x] Commits are signed-off using `git commit -s`
- [x] Automated tests added to cover the changes. If tests couldn't be added, an explanation is provided in the Verification and Testing section.

#### Proposed Changes

Refactor acorn info and support acorn info -A

Lists info for the current default project, or for all projects.

Returns a struct infoCLIResponse, {Client, []ProjectInfo}

Info may be the same for multiple project's, and could be worth while simplifying in the future, however wasn't sure how to go about this while still using a map from Project Name to Info. The only unique item in each info per project is publicKeys.keyID

#### User-Facing Change? ####
acorn info and acorn info -A will have a slightly different response and behavior.

#### Verification and Testing ####

Refactored CLI tests so info_test.go uses gomock. This tests regular info response, no info response, and info -A response.

Test file for testing client and multiclient's implementation of info as well. _integration/client/info/info_test.go_

#### Linked Issues ####
#1197

### CODE REVIEW

File: pkg/client/multi.go
Comment (StrongMonkey) [inline pkg/client/multi.go:430]: It would be cleaner to just do 

```
for _, info := range subprojectInfo {
  info.name = project.name
  projectsInfo = append(projectsInfo, info)
}
```

So if a project can't list info, I fee like it is better UX to users that we just silently ignore it instead of throwing an error.

File: pkg/cli/info_test.go
Comment (StrongMonkey) [inline pkg/cli/info_test.go:54]: You should be able to implement this to get test running 
https://github.com/acorn-io/acorn/blob/34eb2440a787ff78b0717bf1a479c678e75658d4/pkg/cli/testdata/MockClient.go#L614

File: pkg/mocks/Makefile
Comment (cjellick) [inline pkg/mocks/Makefile:1]: Do we need this here and in the top level Makefile? Seems simpler to maintain it in a single place

Comment (jacobdonenfeld) [reply pkg/mocks/Makefile:1]: Good call. Will move to top level

**Comment by cjellick** [comment]
Just so we don't lose track of the feedback: I ran this locally and `acorn info -A` was resulting in:
```
  ✗  ERROR:  invalid project name [acorn.io/cjellick/acorn/acorn.io/cjellick/blah]: must contain zero, one or two slashes [/]
```
this needs addressed before it can be merged. **I'd also like to see if you can add a test that triggers this bug.**

Root cause of this bug is the implementation of `func (m *MultiClient) Info()` and specifically it's calling of 
```projects, err := m.ProjectList(ctx)```
That does not cause the desired effect.

A bit of debugging and inspection indicated that we probably need to use the `aggregate()` function to call Info() for each `client`, similar to how the AppList (or any of the list functions) works:
```
func (m *MultiClient) AppList(ctx context.Context) ([]apiv1.App, error) {
	return aggregate(ctx, m.Factory, func(client Client) ([]apiv1.App, error) {
		return client.AppList(ctx)
	})
}
```
...but the problem is that the aggregate function is going to prepend the project name to the name of whatever resource is returned by the client call. In this case, the object being returned is the Info object and its name IS the project name, so we need a modified version of the aggregate function that doesn't prepend the project name. I suggested the following changes but they actually don't compile because of some funkiness with generics. If we just want to add a bool parameter directly to the `aggregate()` function , rather than trying to keep its signature unchanged, I think that would be fine.
```
diff --git a/pkg/client/multi.go b/pkg/client/multi.go
index b4eaf4d1..7333a24d 100644
--- a/pkg/client/multi.go
+++ b/pkg/client/multi.go
@@ -37,6 +37,10 @@ type ObjectPointer[T any] interface {
 }

 func aggregate[T any, V ObjectPointer[T]](ctx context.Context, factory ProjectClientFactory, cb func(client Client) ([]T, error)) ([]T, error) {
+	return aggregateNameOption(ctx, true, factory, cb)
+}
+
+func aggregateNameOption[T any, V ObjectPointer[T]](ctx context.Context, appendProjectName bool, factory ProjectClientFactory, cb func(client Client) ([]T, error)) ([]T, error) {
 	var result []T
 	clients, err := factory.List(ctx)
 	if err != nil {
@@ -50,7 +54,9 @@ func aggregate[T any, V ObjectPointer[T]](ctx context.Context, factory ProjectCl
 		for _, item := range items {
 			if client.GetProject() != factory.DefaultProject() {
 				p := (V)(&item)
-				p.SetName(client.GetProject() + "/" + p.GetName())
+				if appendProjectName {
+					p.SetName(client.GetProject() + "/" + p.GetName())
+				}
 			}
 			result = append(result, item)
 		}
@@ -409,12 +415,16 @@ func (m *MultiClient) ProjectList(ctx context.Context) ([]apiv1.Project, error)
 }

 func (m *MultiClient) Info(ctx context.Context) ([]apiv1.Info, error) {
-	var projectsInfo []apiv1.Info
+	return aggregateNameOption(ctx, false, m.Factory, func(client Client) ([]apiv1.Info, error) {
+		return client.Info(ctx)
+	})
+
 	// Can this be refactored into a call to aggregate?
 	projects, err := m.ProjectList(ctx)
 	if err != nil {
 		return nil, err
 	}
+
 	var subprojectInfo []apiv1.Info
 	// For each project after listing them all, call info on each and populate projectsInfo slice
 	for _, project := range projects {
```

**Review by cjellick — DISMISSED** [review]
See my latest comment. This will probably be finished while I'm out, so @thedadams and @tylerslaton please re-review this, no need to wait for me to re-review

File: pkg/client/client.go
Comment (cjellick) [reply pkg/client/client.go:235]: Well, the problem is that in Hub server information can vary between projects. Right now, because of how we do upgrades, the server versions happen to be consistent, but we don't want to rely on that.

File: pkg/config/cliconfig.go
Comment (jacobdonenfeld) [reply pkg/config/cliconfig.go:45]: Removed, and moved the other to generate.go. TY

File: pkg/cli/info.go
Comment (jacobdonenfeld) [reply pkg/cli/info.go:35]: Server contains some public-key key/value pair unique to the project.

File: pkg/client/app.go
Comment (thedadams) [inline pkg/client/app.go:66]: Remove comment.

File: pkg/client/multi.go
Comment (thedadams) [inline pkg/client/multi.go:452]: Remove if not necessary.

File: pkg/client/multi.go
Comment (thedadams) [inline pkg/client/multi.go:59]: nit: combine the if-statments:
```
if client.GetProject() != factory.DefaultProject() && appendProjectName {
```

File: integration/client/info/info_test.go
Comment (thedadams) [inline integration/client/info/info_test.go:75]: Remove if not necessary.

File: integration/helper/namespace.go
Comment (thedadams) [inline integration/helper/namespace.go:35]: This looks like a note to yourself. Remove if unnecessary.

File: integration/helper/namespace.go
Comment (thedadams) [inline integration/helper/namespace.go:48]: Namespaces aren't namespaced. I think the api-server will just remove this on creation.

File: integration/helper/namespace.go
Comment (thedadams) [inline integration/helper/namespace.go:57]: nit: I like using `t.Log` here instead of `fmt.Printf` because `t.Log` will only display logs if the test fails or if `-test.v` flag is passed.

**Review by thedadams — APPROVED** [review]
Looks good! Just a few extraneous comments to be removed.

File: integration/helper/namespace.go
Comment (tylerslaton) [inline integration/helper/namespace.go:50]: What if you just took in `metav1.ObjectMeta` here and applied it to the namespace when you create it? That would remove the name, project and generate parameters as well as some of this logic. It also has the benefit of being flexible should we need to add different labels or annotations.

File: integration/helper/namespace.go
Comment (tylerslaton) [inline integration/helper/namespace.go:59]: Could you make this comment a bit more descriptive?

File: pkg/client/client.go
Comment (tylerslaton) [inline pkg/client/client.go:16]: Why is this imported here? Is it so that package's init function runs?

**Review by tylerslaton — APPROVED** [review]
Looks good, I agree with everything that Donnie said and had a few extra comments.

File: pkg/client/client.go
Comment (jacobdonenfeld) [reply pkg/client/client.go:16]: good catch, removed

File: integration/helper/namespace.go
Comment (jacobdonenfeld) [reply integration/helper/namespace.go]: Great point, that cleaned it up a bit. Thank you!

File: integration/secrets/secret_test.go
Comment (jsilverio22) [inline integration/secrets/secret_test.go:142]: will this always have something at info[0]?

File: integration/secrets/secret_test.go
Comment (jsilverio22) [inline integration/secrets/secret_test.go:149]: same as above

**Review by jsilverio22 — APPROVED** [review]
couple quick questions other than that lgtm

File: integration/secrets/secret_test.go
Comment (jacobdonenfeld) [reply integration/secrets/secret_test.go:142]: Either something at [0], or nothing, depending on hub implemention. Right now it always has [0]

**Pull request merged**
